### ISSUE TITLE
Building on Windows 10

### ISSUE THREAD

**Issue by daenuprobst** [ogdf/ogdf#47]
Creating a new `MultilevelGraph `from `GraphAttributes`

```
EdgeWeightedGraph<float> g;
// ...
GraphAttributes graph_attributes(g);
MultilevelGraph mlg(graph_attributes);
```
fails with

> Exception thrown: write access violation.
> ogdf::NodeArray<ogdf::NodeElement * __ptr64>::operator[](...) returned nullptr.

at

https://github.com/ogdf/ogdf/blob/8a103cf3a7dfff87fe8b7534575604bc53c0870c/src/ogdf/energybased/multilevel_mixer/MultilevelGraph.cpp#L210

Building on Windows 10 with:

> 15.9.21.664-- Building for: Visual Studio 15 2017
> -- The CXX compiler identification is MSVC 19.16.27031.1

`cmake .. -DCMAKE_GENERATOR_PLATFORM=x64 -DCMAKE_BUILD_TYPE=Debug -DBUILD_SHARED_LIBS=ON`

`cmake --build . --config Debug --target install -- /m`

Edit:

There is also the assert that gets triggered in Array.h
https://github.com/ogdf/ogdf/blob/8a103cf3a7dfff87fe8b7534575604bc53c0870c/include/ogdf/basic/Array.h#L297

Edit 2:

When compiling with `-DBUILD_SHARED_LIBS=OFF` and then linking the static lib, the assert error is thrown at:

https://github.com/ogdf/ogdf/blob/8a103cf3a7dfff87fe8b7534575604bc53c0870c/include/ogdf/basic/EdgeArray.h#L178

**Comment by sbeyer** [comment]
Thank you for reporting the issue. Can you provide a full minimal example that triggers the error?

To exclude possible compilation mistakes, I recommend that you do it in a file like `doc/examples/basic/issue-47.cpp` (don't forget to re-run CMake after creating) ... then it is automatically compiled with the correct compilation flags when you compile the examples; the executable will be named `doc/examples/basic/ex-issue-47.exe`.

A side note: the `EdgeWeightedGraph` class is more or less a newly-introduced-yet-legacy class that should not be used directly (except when there is no other choice, e.g., when using the Steiner tree algorithms). You may of course use it, if it does what you want. However, since you are using `GraphAttributes` anyway, it may be a better choice to activate the `edgeDoubleWeight` attribute and use the `doubleWeight()` accessor methods.

**Comment by daenuprobst** [comment]
Thanks for the info. I'll probably get back to it next week (currently busy with manuscript writing ;-)).

As for the `EdgeWeightedGraph`, I fairly recently switched to it from `Graph` when I was digging around in the code. So I'll change that again. Thank you.

**Comment by daenuprobst** [comment]
So. sorry for the delay, but holidays and paper-writing got in the way. Here is a minimal-ish example that generates an error before reaching any of the input. I assume these problems are connected:

```python
std::cout << "Testing ..." << std::endl;

uint32_t vertex_count = 5;
std::vector<std::tuple<uint32_t, uint32_t, float>> edges;
edges.emplace_back(std::make_tuple(0, 1, 0.5));
edges.emplace_back(std::make_tuple(1, 2, 0.5));
edges.emplace_back(std::make_tuple(2, 3, 0.5));
edges.emplace_back(std::make_tuple(3, 4, 0.5));
edges.emplace_back(std::make_tuple(4, 0, 0.5));
edges.emplace_back(std::make_tuple(2, 4, 2.0));

tmap::GraphProperties gp;

Graph g;
GraphAttributes ga(g, GraphAttributes::nodeGraphics|GraphAttributes::edgeGraphics|GraphAttributes::edgeDoubleWeight);

std::vector<std::vector<uint32_t>> adjacency_list(vertex_count);
std::vector<uint32_t> degrees(vertex_count);
std::vector<node> index_to_node(vertex_count);

std::cout << "Adding nodes ..." << std::endl;

for (uint32_t i = 0; i < vertex_count; i++)
    index_to_node[i] = g.newNode();

std::cout << "Adding edges ..." << std::endl;

for (size_t i = 0; i < edges.size(); i++) {
    edge e = g.newEdge(index_to_node[std::get<0>(edges[i])], index_to_node[std::get<1>(edges[i])]);
    std::cout << index_to_node[std::get<0>(edges[i])] << std::endl;
    std::cout << index_to_node[std::get<1>(edges[i])] << std::endl;
    std::cout << e << std::endl;
    std::cout << e->graphOf() << std::endl;
    ga.doubleWeight(e) = std::get<2>(edges[i]);
}

std::cout << "Added edges ..." << std::endl;

tmap::LayoutConfiguration config;

LayoutInternal(g, ga, vertex_count, config, gp);
std::cout << "Done ..." << std::endl;
```

The output of this project is the following:
```
Testing ...
Adding nodes ...
Adding edges ...
0
1
(0,1)
0000000000000001
Assertion failed: e->graphOf() == m_pGraph, file C:\Users\...\Miniconda3\conda-bld\tmap_1565530533355\_build_env\include\ogdf/basic/EdgeArray.h, line 17
```

I am currently trying to build the same project that built on Linux with gcc and clang on macOS, where I encounter the same problems (with gcc-9). As you see from the output, it is a Python project, but this shouldn't be a problem--I guess. I'll keep investigating...

**Comment by daenuprobst** [comment]
I decoupled this example from Python and built it as part of the build-all target. (Quick and dirty: I just replaced the code of one of the examples) and the code ran through without a problem. The output is the following:
```
Testing ...
Adding nodes ...
Adding edges ...
0
1
(0,1)
000000082435F310
1
2
(1,2)
000000082435F310
2
3
(2,3)
000000082435F310
3
4
(3,4)
000000082435F310
4
0
(4,0)
000000082435F310
2
4
(2,4)
000000082435F310
Added edges ...
Done ...
```
It almost looks like there is some memory weirdness going on when it is run on Windows / macOS as a Python extension... I'll have to look into this. Any pointers?

**Comment by sbeyer** [comment]
> It almost looks like there is some memory weirdness going on when it is run on Windows / macOS as a Python extension... I'll have to look into this. Any pointers?

Yes. Most probably the library is compiled in release mode and your program in debug mode, or vice versa. Some basic classes are defined differently in debug/release mode, so addresses to class attributes are totally f**ked up and OGDF code does only weird things.

**Comment by daenuprobst** [comment]
Thanks for the swift answer. I checked whether there is any release / debug mismatch and could rule it out. For now, I set aside. I tried a few other things with no luck. However, I created a minimal-ish example that can be found [here](https://cloud.gdb.tools/s/cJwTdFAL8r4dMZT) (I put it on our research groups cloud on the Uni Bern). I actually split the project into two parts. 

- One is a conda package of OGDF, which is actually already available via `conda install -c tmap ogdf`. I created it as a package with the goal to also provide others with an easy-to-get-started  OGDF distribution.
- The other is a (stripped down) conda package of one of our recent projects that relies on OGDF (the preprint should be up during the next 48 hours).

So that's the background. As I said, it works perfectly well on Linux. So today I spent trying to build it on Windows without the errors mentioned above. I will continue trying to get it to work, but I'm slowly running out of ideas, except maybe replace pybind11 with boost.python.

**Comment by sbeyer** [comment]
> Thanks for the swift answer. I checked whether there is any release / debug mismatch and could rule it out.

How? Have you tried to compile [check-build-mode.cpp](../blob/master/doc/examples/special/check-build-mode.cpp) with your configuration?

**Comment by daenuprobst** [comment]
I just double and triple checked everything, but check-build-mode.cpp seems like a better option and I will try it on my Windows machine. However, I now got the macOS build working, where the solution turned out to be to use the compiler provided by conda / anaconda (this is obviously a problem on Windows where they cannot redistribute MSVC). As an aside, I made to project and the associated paper public (see the [project page](http://tmap.gdb.tools) and the associated [paper](https://chemrxiv.org/articles/Visualization_of_Very_Large_High-Dimensional_Data_Sets_as_Minimum_Spanning_Trees/9698861)), let me know how you would like OGDF be linked on the project website.

**Comment by sbeyer** [comment]
> let me know how you would like OGDF be linked on the project website.

If you want to link to OGDF, just use ogdf.net

**Comment by milsen** [comment]
I'll close this since it's a very old issue concerning an old OGDF version; and at least a partial solution seems to have been found anyway.
